sql: add information about retries to statement trace

xinhaoz opened the pull request:
Currently, statement diagnostic bundles reflect the last,
successful run of the query. While we already track the number
of times a transaciton has been automatically retried, we do
not record errors causing the retry. This addition allows
us to access the last error causing an automatic retry
during statement execution, adding it to the statement
trace and diagnostic bundle.

Resolves: cockroachdb#65146

Release note (sql change):  Retry information has been added
to the statement trace under the 'exec stmt' operation.

The trace message is in the format:
"executing after <int> retries, last retry reason: <err>"

This message will appear in any operations that show the
statement trace, which is included in operations such as
SHOW TRACE FOR SESSION and is also exported in the statement
diagnostics bundle.

--------

Sample message in the trace after forcing a retry:
```
"sql/conn_executor_exec.go:690 [n1,client=127.0.0.1:54187,hostnossl,user=root] executing after 4 retries, last retry reason: crdb_internal.force_retry(): TransactionRetryWithProtoRefreshError: forced by crdb_internal.force_retry()"
```

xinhaoz committed:
sql: record last auto retryable error to extraTxnState

Currently, statement diagnostic bundles reflect the last,
successful run of the query. While we already track the number
of times a transaciton has been automatically retried, we do
not record errors causing the retry. This addition allows
us to access the last error causing an automatic retry
during statement execution, adding it to the statement
trace and diagnostic bundle.

Resolves: #65146

Release note (sql change): Retry information has been added
to the statement trace under the 'exec stmt' operation.

The trace message is in the format:
"executing after <int> retries, last retry reason: <err>"

This message will appear in any operations that show the
statement trace, which is included in operations such as
SHOW TRACE FOR SESSION and is also exported in the statement
diagnostics bundle.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67941)
<!-- Reviewable:end -->

maryliag reviewed (commented):
Great job on this!

A few things to add to your PR/commit message:
- the issue this is related to with `Resolves: #issuenumber`
- an example of a result, showing the output of your changes
- you're making a change that the user will notice, so it should be a release note for it, you can add a message mentioning which new information in being added and where

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-:-MfJyIwICtX7oG3_EZn-:b8borrd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @matthewtodd, and @THardy98)



<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67941)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd, @THardy98, and @xinhaoz)

___
*[pkg/sql/txn_restart_test.go, line 1598 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfZGYdv8l0Pz5yuQIqL:-MfZGYdv8l0Pz5yuQIqM:b-kyd6jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f70692ea83be8413571c673babb093857a075d9a/pkg/sql/txn_restart_test.go#L1598)):*
> ```Go
> 	params.Knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
> 		BeforeRestart: func(ctx context.Context, reason error) {
> 			if !testutils.IsError(reason, fmt.Sprintf("injected err %d", retryCount)) {
> ```

nit: IIUC, if somehow the txn is not restarted at all for some reason, this test will pass. Can we add a check to ensure that the callback is actually executed? 
___
*[pkg/sql/txn_restart_test.go, line 1622 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfZFnQq9Wzb1iHaRtl5:-MfZFnQq9Wzb1iHaRtl6:b-rqso30) ([raw file](https://github.com/cockroachdb/cockroach/blob/f70692ea83be8413571c673babb093857a075d9a/pkg/sql/txn_restart_test.go#L1622)):*
> ```Go
> 	defer cleanupFilter()
> 
> 	sqlDB.SetMaxOpenConns(1)
> ```

Why do we need to set max open connection here?
___
*[pkg/sql/txn_restart_test.go, line 1624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfZFI1x92VkHtqat1f8:-MfZFI1x92VkHtqat1f9:bjztl2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f70692ea83be8413571c673babb093857a075d9a/pkg/sql/txn_restart_test.go#L1624)):*
> ```Go
> 	sqlDB.SetMaxOpenConns(1)
> 
> 	if _, err := sqlDB.Exec(`
> ```

nit: you can wrap `sqlDB` into a `sqlutils.MakeSQLRunner` so it will fail the test if there's any error encountered. 


<!-- Sent from Reviewable.io -->

xinhaoz commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67941)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd, @THardy98, and @xinhaoz)
> 
> _[pkg/sql/txn_restart_test.go, line 1598 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfZGYdv8l0Pz5yuQIqL:-MfZGYdv8l0Pz5yuQIqM:b-kyd6jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f70692ea83be8413571c673babb093857a075d9a/pkg/sql/txn_restart_test.go#L1598)):_
> 
> > ```go
> > 	params.Knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
> > 		BeforeRestart: func(ctx context.Context, reason error) {
> > 			if !testutils.IsError(reason, fmt.Sprintf("injected err %d", retryCount)) {
> > ```
> 
> nit: IIUC, if somehow the txn is not restarted at all for some reason, this test will pass. Can we add a check to ensure that the callback is actually executed?
> 
> _[pkg/sql/txn_restart_test.go, line 1622 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfZFnQq9Wzb1iHaRtl5:-MfZFnQq9Wzb1iHaRtl6:b-rqso30) ([raw file](https://github.com/cockroachdb/cockroach/blob/f70692ea83be8413571c673babb093857a075d9a/pkg/sql/txn_restart_test.go#L1622)):_
> 
> > ```go
> > 	defer cleanupFilter()
> > 
> > 	sqlDB.SetMaxOpenConns(1)
> > ```
> 
> Why do we need to set max open connection here?
> 
> _[pkg/sql/txn_restart_test.go, line 1624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfZFI1x92VkHtqat1f8:-MfZFI1x92VkHtqat1f9:bjztl2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f70692ea83be8413571c673babb093857a075d9a/pkg/sql/txn_restart_test.go#L1624)):_
> 
> > ```go
> > 	sqlDB.SetMaxOpenConns(1)
> > 
> > 	if _, err := sqlDB.Exec(`
> > ```
> 
> nit: you can wrap `sqlDB` into a `sqlutils.MakeSQLRunner` so it will fail the test if there's any error encountered.

Added restarted check and SQL runner. The `SetMaxOpenConns` snuck in on accident, removed it as it shouldn't be necessary for this test.

Azhng reviewed (commented):
Nice. The code looks good. 

@yuzefovich  do you mind take a look as well? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-:-MfcYJsy961gmipdqY5Y:b-i8f081)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, @THardy98, and @xinhaoz)

___
*[pkg/sql/conn_executor.go, line 631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfcXiGIChkam2JZk4jR:-MfcXiGIChkam2JZk4jS:b-qlewb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bb6b3ddd698a266799406f36639a2c0c70b48d4/pkg/sql/conn_executor.go#L631)):*
> ```Go
> }
> 
> // GetLocalIndexStatistics returns a idxusage.LocalIndexUsageStats.
> ```

This seems to be included by accident ? 
___
*[pkg/sql/txn_restart_test.go, line 1631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfcYB0f3NcHdfSaGwjG:-MfcYB0f3NcHdfSaGwjH:bcqzib4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bb6b3ddd698a266799406f36639a2c0c70b48d4/pkg/sql/txn_restart_test.go#L1631)):*
> ```Go
> `)
> 
> 	if retryCount != numRetries {
> ```

super nit: you can use `require.Equal()` method here to reduce some verbosity


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice, the code looks good to me too.

> Reviewed 1 of 1 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-:-MfctBmt568hyhO6ric-:b-m1pgm2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, @matthewtodd, @THardy98, and @xinhaoz)

___
*[pkg/sql/conn_executor.go, line 1210 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-MfcsWCu25OLmSqNjd1c:-MfcsWCu25OLmSqNjd1d:blqu3rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bb6b3ddd698a266799406f36639a2c0c70b48d4/pkg/sql/conn_executor.go#L1210)):*
> ```Go
> 		// the current transaction is being automatically retried. This is used in
> 		// statement traces to give more information in statement diagnostic bundles.
> 		autoRetryReason error
> ```

super nit: might be nicer to put this right below `autoRetryCounter`.


<!-- Sent from Reviewable.io -->

Azhng reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :fire: :fire: :fire:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67941#-:-MfdDLcQCQQJ3-MS3boA:be56upl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @maryliag, @matthewtodd, @THardy98, @xinhaoz, and @yuzefovich)



<!-- Sent from Reviewable.io -->

xinhaoz commented:
bors r+
